### PR TITLE
fix a bug when bagging with reset_config

### LINKED ISSUE #2144: Check failed: bag_cnt == num_data_ at serial_tree_learner
The below error occurs when I run LGB on different machines. I guess it may be related to my dataset since when I change the dataset the error message disappears. However, it is still strange because when I remove either the `bagging_fraction` parameter or the learning rate decay callback, the model can be trained successfully. 

## Environment info

Operating System: Linux Centos

CPU/GPU model: CPU Intel(R) Xeon(R) Gold 61xx CPU

C++/Python/R version: Python 3.6

LightGBM version or commit hash: 2.2.3

## Error message

```
---------------------------------------------------------------------------
LightGBMError                             Traceback (most recent call last)
<ipython-input-103-545a3739dbbe> in <module>()
     10 }
     11 train = lgb.Dataset(train_data[names], label=train_data['y'], free_raw_data=False, )
---> 12 bst = lgb.train(param, train, num_round, callbacks=[learning_rate_decay])

/usr/local/lib/python3.6/dist-packages/lightgbm/engine.py in train(params, train_set, num_boost_round, valid_sets, valid_names, fobj, feval, init_model, feature_name, categorical_feature, early_stopping_rounds, evals_result, verbose_eval, learning_rates, keep_training_booster, callbacks)
    216                                     evaluation_result_list=None))
    217 
--> 218         booster.update(fobj=fobj)
    219 
    220         evaluation_result_list = []

/usr/local/lib/python3.6/dist-packages/lightgbm/basic.py in update(self, train_set, fobj)
   1800             _safe_call(_LIB.LGBM_BoosterUpdateOneIter(
   1801                 self.handle,
-> 1802                 ctypes.byref(is_finished)))
   1803             self.__is_predicted_cur_iter = [False for _ in range_(self.__num_dataset)]
   1804             return is_finished.value == 1

/usr/local/lib/python3.6/dist-packages/lightgbm/basic.py in _safe_call(ret)
     44     """
     45     if ret != 0:
---> 46         raise LightGBMError(decode_string(_LIB.LGBM_GetLastError()))
     47 
     48 

LightGBMError: Check failed: bag_cnt == num_data_ at /__w/1/s/python-package/compile/src/treelearner/serial_tree_learner.cpp, line 786 .
```

## Reproducible examples

```
import numpy
import lightgbm as lgb
import pandas as pd

# load dataset
train_data = pd.read_hdf('train_data.h5', key='df')   

# Parameters
num_round = 3000
param = {
    'learning_rate': 0.05,
    'num_leaves': 25,
    'max_depth': 7,
    'bagging_freq': 5,
    'bagging_fraction': 0.8,
    
    'objective': 'mae',
    'metric': ['mae'],
}

learning_rate_decay = lgb.reset_parameter(learning_rate=lambda iter: np.max([0.05 * (0.9998 ** iter), 0.001]))

train = lgb.Dataset(train_data.loc[:,train_data.columns != 'y'], label=train_data['y'], free_raw_data=False)
bst = lgb.train(param, train, num_round, callbacks=[learning_rate_decay])
```

## Steps to reproduce

1. Download the [dataset](https://github.com/jaydu1/img/raw/master/train_data_.h5)
2. Run the above script

### PR DESCRIPTION
**Pull request by guolinke**
to fix #2144, and some code cleans.

### CODE REVIEW

File: src/boosting/gbdt.cpp [hunk 1]
```diff
@@ -688,6 +688,11 @@ void GBDT::ResetConfig(const Config* config) {
 void GBDT::ResetBaggingConfig(const Config* config, bool is_change_dataset) {
   // if need bagging, create buffer
   if (config->bagging_fraction < 1.0 && config->bagging_freq > 0) {
+    need_re_bagging_ = false;
+    if (!is_change_dataset &&
+      config_.get() != nullptr && config_->bagging_fraction == config->bagging_fraction && config_->bagging_freq == config->bagging_freq) {
+      return;
+    }
```
Comment (guolinke) [inline src/boosting/gbdt.cpp]: do nothing if no bagging config changes.

File: src/boosting/gbdt.cpp [hunk 2]
```diff
@@ -719,9 +724,7 @@ void GBDT::ResetBaggingConfig(const Config* config, bool is_change_dataset) {
       Log::Debug("Use subset for bagging");
     }
 
-    if (is_change_dataset) {
-      need_re_bagging_ = true;
-    }
+    need_re_bagging_ = true;
```
Comment (guolinke) [inline src/boosting/gbdt.cpp]: need to rebagging when bagging_config changes

File: src/boosting/rf.hpp [hunk 3]
```diff
@@ -130,7 +130,8 @@ class RF : public GBDT {
       }
 
       if (new_tree->num_leaves() > 1) {
-        tree_learner_->RenewTreeOutput(new_tree.get(), objective_function_, init_scores_[cur_tree_id],
+        std::vector<double> tmp_preds(num_data_, init_scores_[cur_tree_id]);
+        tree_learner_->RenewTreeOutput(new_tree.get(), objective_function_, tmp_preds.data(),
```
Comment (guolinke) [inline src/boosting/rf.hpp]: not efficient, but this can reduce many duplicated codes.

**Comment by StrikerRUS** [comment]
@guolinke This PR introduced some `-Wunused-parameter` warnings. Please take a look.

```
Executing transaction: ...working... done
-- The C compiler identification is GNU 4.8.4
-- The CXX compiler identification is GNU 4.8.4
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found OpenMP_C: -fopenmp (found version "3.1") 
-- Found OpenMP_CXX: -fopenmp (found version "3.1") 
-- Found OpenMP: TRUE (found version "3.1")  
-- Configuring done
-- Generating done
-- Build files have been written to: /__w/1/s/build
Scanning dependencies of target _lightgbm
[  3%] Building CXX object CMakeFiles/_lightgbm.dir/src/c_api.cpp.o
[  6%] Building CXX object CMakeFiles/_lightgbm.dir/src/application/application.cpp.o
[  9%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/boosting.cpp.o
[ 12%] Building CXX object CMakeFiles/_lightgbm.dir/src/lightgbm_R.cpp.o
In file included from /__w/1/s/src/boosting/gbdt.h:9:0,
                 from /__w/1/s/src/boosting/dart.hpp:16,
                 from /__w/1/s/src/boosting/boosting.cpp:7:
/__w/1/s/include/LightGBM/objective_function.h:46:18: warning: unused parameter ‘residual_getter’ [-Wunused-parameter]
   virtual double RenewTreeOutput(double ori_output, std::function<double(const label_t*, int)> residual_getter,
                  ^
In file included from /__w/1/s/include/LightGBM/metric.h:11:0,
                 from /__w/1/s/src/c_api.cpp:11:
/__w/1/s/include/LightGBM/objective_function.h:46:18: warning: unused parameter ‘residual_getter’ [-Wunused-parameter]
   virtual double RenewTreeOutput(double ori_output, std::function<double(const label_t*, int)> residual_getter,
                  ^
In file included from /__w/1/s/include/LightGBM/metric.h:11:0,
                 from /__w/1/s/src/application/application.cpp:10:
/__w/1/s/include/LightGBM/objective_function.h:46:18: warning: unused parameter ‘residual_getter’ [-Wunused-parameter]
   virtual double RenewTreeOutput(double ori_output, std::function<double(const label_t*, int)> residual_getter,
                  ^
[ 15%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/gbdt.cpp.o
In file included from /__w/1/s/src/boosting/gbdt.h:9:0,
                 from /__w/1/s/src/boosting/gbdt.cpp:5:
/__w/1/s/include/LightGBM/objective_function.h:46:18: warning: unused parameter ‘residual_getter’ [-Wunused-parameter]
   virtual double RenewTreeOutput(double ori_output, std::function<double(const label_t*, int)> residual_getter,
                  ^
cc1plus: warning: unrecognized command line option "-Wno-ignored-attributes" [enabled by default]
cc1plus: warning: unrecognized command line option "-Wno-ignored-attributes" [enabled by default]
[ 18%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/gbdt_model_text.cpp.o
[ 21%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/gbdt_prediction.cpp.o
In file included from /__w/1/s/include/LightGBM/metric.h:11:0,
                 from /__w/1/s/src/boosting/gbdt_model_text.cpp:5:
/__w/1/s/include/LightGBM/objective_function.h:46:18: warning: unused parameter ‘residual_getter’ [-Wunused-parameter]
   virtual double RenewTreeOutput(double ori_output, std::function<double(const label_t*, int)> residual_getter,
                  ^
In file included from /__w/1/s/src/boosting/gbdt_prediction.cpp:5:0:
/__w/1/s/include/LightGBM/objective_function.h:46:18: warning: unused parameter ‘residual_getter’ [-Wunused-parameter]
   virtual double RenewTreeOutput(double ori_output, std::function<double(const label_t*, int)> residual_getter,
                  ^
cc1plus: warning: unrecognized command line option "-Wno-ignored-attributes" [enabled by default]
[ 25%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/prediction_early_stop.cpp.o
cc1plus: warning: unrecognized command line option "-Wno-ignored-attributes" [enabled by default]
[ 28%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/bin.cpp.o
[ 31%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/config.cpp.o
cc1plus: warning: unrecognized command line option "-Wno-ignored-attributes" [enabled by default]
[ 34%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/config_auto.cpp.o
cc1plus: warning: unrecognized command line option "-Wno-ignored-attributes" [enabled by default]
[ 37%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/dataset.cpp.o
[ 40%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/dataset_loader.cpp.o
[ 43%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/file_io.cpp.o
[ 46%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/json11.cpp.o
[ 50%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/metadata.cpp.o
[ 53%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/parser.cpp.o
[ 56%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/tree.cpp.o
[ 59%] Building CXX object CMakeFiles/_lightgbm.dir/src/metric/dcg_calculator.cpp.o
In file included from /__w/1/s/include/LightGBM/metric.h:11:0,
                 from /__w/1/s/src/metric/dcg_calculator.cpp:5:
/__w/1/s/include/LightGBM/objective_function.h:46:18: warning: unused parameter ‘residual_getter’ [-Wunused-parameter]
   virtual double RenewTreeOutput(double ori_output, std::function<double(const label_t*, int)> residual_getter,
                  ^
[ 62%] Building CXX object CMakeFiles/_lightgbm.dir/src/metric/metric.cpp.o
In file included from /__w/1/s/include/LightGBM/metric.h:11:0,
                 from /__w/1/s/src/metric/metric.cpp:5:
/__w/1/s/include/LightGBM/objective_function.h:46:18: warning: unused parameter ‘residual_getter’ [-Wunused-parameter]
   virtual double RenewTreeOutput(double ori_output, std::function<double(const label_t*, int)> residual_getter,
                  ^
[ 65%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/linker_topo.cpp.o
cc1plus: warning: unrecognized command line option "-Wno-ignored-attributes" [enabled by default]
[ 68%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/linkers_mpi.cpp.o
[ 71%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/linkers_socket.cpp.o
[ 75%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/network.cpp.o
[ 78%] Building CXX object CMakeFiles/_lightgbm.dir/src/objective/objective_function.cpp.o
In file included from /__w/1/s/src/objective/objective_function.cpp:5:0:
/__w/1/s/include/LightGBM/objective_function.h:46:18: warning: unused parameter ‘residual_getter’ [-Wunused-parameter]
   virtual double RenewTreeOutput(double ori_output, std::function<double(const label_t*, int)> residual_getter,
                  ^
[ 81%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/data_parallel_tree_learner.cpp.o
[ 84%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/feature_parallel_tree_learner.cpp.o
[ 87%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/gpu_tree_learner.cpp.o
[ 90%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/serial_tree_learner.cpp.o
In file included from /__w/1/s/src/treelearner/serial_tree_learner.cpp:8:0:
/__w/1/s/include/LightGBM/objective_function.h:46:18: warning: unused parameter ‘residual_getter’ [-Wunused-parameter]
   virtual double RenewTreeOutput(double ori_output, std::function<double(const label_t*, int)> residual_getter,
                  ^
cc1plus: warning: unrecognized command line option "-Wno-ignored-attributes" [enabled by default]
[ 93%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/tree_learner.cpp.o
[ 96%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/voting_parallel_tree_learner.cpp.o
cc1plus: warning: unrecognized command line option "-Wno-ignored-attributes" [enabled by default]
cc1plus: warning: unrecognized command line option "-Wno-ignored-attributes" [enabled by default]
[100%] Linking CXX shared library ../lib_lightgbm.so
[100%] Built target _lightgbm
```

**Comment by guolinke** [comment]
@StrikerRUS thanks ! i will fix it

**Pull request merged**